specs/commandline/blob.md

```diff
@@ -0,0 +1,363 @@
+  -e,  --expiry duration            optional expiry that provides a "best by use" time for the blob. The duration is specified in minutes(m) and/or hours(h). For example: 12h, 30m, 3h20m
+       --id string                  key id (required if --plugin is set). This is mutually exclusive with the --key flag
+  -k,  --key string                 signing key name, for a key previously added to notation's key list. This is mutually exclusive with the --id and --plugin flags
+       --plugin string              signing plugin name. This is mutually exclusive with the --key flag
+       --plugin-config stringArray  {key}={value} pairs that are passed as it is to a plugin, refer plugin's documentation to set appropriate values.
+       --signature-format string    signature envelope format, options: "jws", "cose" (default "jws")
+  -m,  --user-metadata stringArray  {key}={value} pairs that are added to the signature payload
+  -d,  --debug                      debug mode
+  -v,  --verbose                    verbose mode
+  -h,  --help                       help for sign
+```
+
+### notation blob inspect
+
+```text
+Inspect a signature associated with a blob
+
+Usage:
+  notation blob inspect [flags] <signature_path>
+
+Flags:
+  -o, --output string         output format, options: 'json', 'text' (default "text")
+  -d, --debug                 debug mode
+  -v, --verbose               verbose mode
+  -h, --help                  help for inspect
+```
+
+### notation blob verify
+
+```text
+Verify a signature associated with a blob
+
+Usage:
+  notation blob verify [flags] --signature <signature_path> <blob_path>
+
+Flags:
+  -s, --signature path        location of the detached signature
+      --media-type string     optional media type of the blob to verify
+      --policy-scope string   optional policy scope to verify against. If not provided, notation verifies against wildcard policy if it exists.
+  -m, --user-metadata stringArray   user defined {key}={value} pairs that must be present in the signature for successful verification if provided
+  -o, --output string         output format, options: 'json', 'text' (default "text")
+  -d, --debug                 debug mode
+  -v, --verbose               verbose mode
+  -h, --help                  help for inspect
+```
+
+## Usage
+
+## Produce detached blob signatures
+
+### Sign a blob by adding a new key
+
+```shell
+# Prerequisites:
+# - A signing plugin is installed. See plugin documentation (https://github.com/notaryproject/notaryproject/blob/main/specs/plugin-extensibility.md) for more details.
+# - Configure the signing plugin as instructed by plugin vendor.
+
+# Add a default signing key referencing the remote key identifier, and the plugin associated with it.
+notation key add --default --name <key_name> --plugin <plugin_name> --id <remote_key_id>
+
+# sign a blob
+notation blob sign --signature-name my-blob-signature /tmp/my-blob.bin
```

Comment:
That is correct. Is this functionality absolutely needed? Would do you suggest adding a flag for `signature-directory`?

Reply:
I think the original `--signature` is good enough and it is consistent with `notation blob verify`.

Reply:
or `--signature-prefix` can be a candidate if `.sig` is keep.
or `--signature-name` allows folders like `foo/bar/my-blob-signature`.

Reply:
@shizhMSFT 
IMO these two options can be confusing. `--signature-prefix` if fine if it is just accepting the prefix for the signature file name. But if it also accepts a path, then the flag name is not very apparent.
`--signature-name` accepting a path is also bit misleading and error prone.

If we want to support arbitrary paths, I recommend doing this

1. Accept an optional flag `--signature-directory`
2. If it is provided, notation will write the <source file name>+".sig.jws" to that directory. If not, then writes to the currently working directory.
This way, the name of the file is not configurable but the path is. I think this is better than providing two many flags for configuring both name and path. 

LMK your thoughts.

Reply:
I think it would be better to enable users choose which folder they want to generate the signature since this is a common scenario. 
For 2, is it really necessary to make the signature envelop type ".sig.jws" as the file extension name?

Reply:
@rgnote Here is the discussion result from community meeting 11/7/2023

- Notation controls the signature file name, for example, use the file name of the blob file + ".sig.jws". Then by default, the signature is stored on the current directory.
- Provide an option for users to specify the signature directory
- If users sign the blob for the 2nd time, overwrite the existing signature by asking whether users are ok with it, if users are OK, then overwriting the existing signature; if not, users can specify a different directory for storing the signature.
- consider `cp` behavior for output file path resolution. 
        - Should notation create a fully nested path or not? 
        - If its directory then should we outupt {dir}/{pattern}.{ext} 
            - if not, should we output the file to the output path and consider that it's a fully qualified filename.
        - `cp` does not create nested paths so maybe to keep this consistent with other tools and avoid surprises. notation could follow the same pattern of not creating nested paths. 

/cc: @shizhMSFT @Two-Hearts @priteshbandi feel free to provide additional comments.

Reply:
Perfect.

Two comments on the above proposal though.
>  If users sign the blob for the 2nd time, overwrite the existing signature by asking whether users are ok with it, if users are OK, then overwriting the existing signature; if not, users can specify a different directory for storing the signature.

I'm not a big fan interactive CLI tools as they make automation hard to achieve. That being said, we should do this.
1. Let notation fail by default if there is a signature file already
2. Provide an optional `--overwrite` flag for suppressing the above failure

> consider cp behavior for output file path resolution.

I think we should keep this behavior simple. Notation being smarter to proactively create non-existing directories or writing signatures directly when the path is not a directory, etc will only add sharp edges. I would say, let's start with the basic functionality i.e. fail if the provided path is not a directory or does not exist. This is not a one-way door. If there is enough demand for handling non-existing directories or configuring signature file name, then we can add them later.

Reply:
@rgnote  I am OK with it. @shizhMSFT @patrickzheng200 @priteshbandi any comments?

Reply:
For overwriting signature we can support two mode:
1. Without `--overwrite` flag; ask for confirmation from user
2. With `--overwrite` flag; overwrite file without any confirmation prompt.

Reply:
Decided on the notary call to go with `--force`, to be consisted with rest of Notation UX.